Zend/zend_lazy_objects.c

```diff
@@ -278,9 +303,17 @@ ZEND_API zend_object *zend_object_make_lazy(zend_object *obj,
 
 		GC_DEL_FLAGS(obj, IS_OBJ_DESTRUCTOR_CALLED);
 
-		/* unset() dynamic properties */
-		zend_object_dtor_dynamic_properties(obj);
```

Comment:
I don't understand why this change is necessary. Reverting it to the previous code doesn't hit a test failure.

Reply:
This is not fully related to foreach, and I should probably commit this change separately, but I realized that setting properties to null here could break some assumptions. For instance, we have code like this that would break if we did it:

```
if (!obj->properties) {
    rebuild_properties(obj);
}
do_something();
obj->properties->..
```

So, just removing the dynamic properties from the ht is safer.

Reply:
Wouldn't that also mean you're potentially accessing the wrong properties?

Reply:
Yes, if do_something() resets the object, the last line could potentially access a property of the object after it has been reset (the right property in the same object, but it has been reset). In the worse case this would allow to observe the state of a non-initialized lazy object.

I've made this change just in case, but it should not happen in practice because this kind of code is unsafe (before lazy objects). Normally we use the result of rebuild_properties() / get_properties() immediately without effects in between (except in # 15938 and maybe in ArrayObject). Outside of the VM and object handlers we use get_properies_for(), which doesn't have this problem because it increases the refcount.